README
Introduction
------------
This is a Tcl extension utilising the libmagic to provide scriptlevel access to
determine the type of a file or buffer. It either returns a descriptive text or
the mime type and optionally encoding.

License
-------
The tmag extension code is copyrighted by
Matthias Kraft <M.Kraft@gmx.com>.

It is based on Tcl's sampleextension which is copyrighted by
Scriptics Corporation, ActiveState and others.

This package is a freely available open source package. You can do virtually
anything you like with it, such as modifying it, redistributing it, and selling
it either in whole or in part.

See the file "license.terms" for complete information.

Building
--------
Building under most UNIX systems is easy, just run the configure script and then
run make. For more information about the build process, see the tcl/unix/README
file in the Tcl src dist. The following minimal example will install the
extension in the /opt/tcl directory.

	$ cd tmag
	$ mkdir build
	$ cd build
	$ ../configure --prefix=/opt/tcl --with-tcl=/opt/tcl/lib
	$ make
	$ make install

You will need to have the Tcl development files (headers and tclConfig script)
installed. Same thing for libmagic, magic.h must be available to the compiler.

Usage
-----

Once built and installed it can be used like this in a script:

    # load extension
    package require tmag

    # identify a file's type with a human readable discription
    set description [libmagic::filetype $file]

    # identify a buffer's content returning a machine readable mimetype
    set mimetype [libmagic::mimetype $buffer -isbuffer]

The prototypes of both commands are:

    libmagic::filetype filename ?(-isbuffer|-follow)? ?-all? ?-raw?

    libmagic::mimetype filename ?(-isbuffer|-follow)? ?-all? ?-with-encoding?

Where 'filename' is either a path to an existing file, or if '-isbuffer' is set
it is a bytearray with the content to identify. The remaining flags are:

    -follow: if set and the filename points to a symlink, resolve the symlink.
             Note: -follow and -isbuffer are mutual exclusive.

    -all: if the content can be identified by more than one type, return all,
          the first match otherwise.

    -raw: do not escape non-printable characters in result text

    -with-encoding: append a semi-colon and a string identifying the encoding
                    of the content.

--
Matthias Kraft

ChangeLog
2012-07-14  Matthias Kraft  <M.Kraft@gmx.com>

	generic/tmag.c: Fixed violation of T_GIFO contract. Thanks go to
	Andreas Kupries for the patch and Brian Griffin for raising the awareness.
	Also closing a case statement properly now and fixing a couple of compiler
	warnings. Gives a proper error message for an unload attempt.

	configure configure.in: Fixed removing source of man page, if built in root
	dir. Bumped patchlevel. Regenerated using newer autoconfig.

	tests/tmag.test: Updated a couple of test cases to follow recent fixes and
	changes of libmagic output. Tested against file-5.11 package.

2009-03-21  Matthias Kraft  <M.Kraft@gmx.com>

	README doc/tmag.man tests/tmag.test: adapted tests and
	documentation to changed output of file-5.00 release.

	generic/tmag.c tests/tmag.test: fixed bug in reporting of system
	errors.

2009-03-20  Matthias Kraft  <M.Kraft@gmx.com>

	configure configure.in Makefile.in pkgIndex.tcl.in tclconfig/*:
	updated to TEA 3.7 from tcl-sampleextension repository.

2009-03-09  Matthias Kraft  <M.Kraft@gmx.com>

	configure configure.in Makefile.in: bumped version to 0.5; added
	--with-magic and --with-dtplite configure options; fixed Makefile
	to htmldir for HTML documentation, to install README and license,
	and to remove certain files

2009-03-06  Matthias Kraft  <M.Kraft@gmx.com>

	configure configure.in Makefile.in README config.man.in
	doc/tmag.man generic/tmag.c generic/tmag.h: made raw output mode
	configurable; now generating and deploying HTML documentation,
	too; changed the ways how the version information is creeping into
	the results;

	tests/tmag.test tests/pwrdLogo75.{gif,jpg,png,ppm,xpm}
	tests/test.{pdf,ps}: added more test cases with some real data.

	configure configure.in: bumped version to 0.4

2009-02-14  Matthias Kraft  <M.Kraft@gmx.com>

	doc/tmag.man.in: documentation in doctools format

	configure configure.in Makefile.in: updated build for documentation

	generic/tmag.c: fixed -all to return raw output

2008-12-07  Matthias Kraft  <M.Kraft@gmx.com>

	README: provisional short documentation

	tests/tmag.test: new file with a couple of test cases

2008-12-06  Matthias Kraft  <M.Kraft@gmx.com>

	generic/tmag.c generic/tmag.h.in: code cleanup, refactoring,
	added runtime configuration and buffer type identification.

	configure configure.in: bumped version

2008-11-02  Matthias Kraft  <M.Kraft@gmx.com>

	*** Imported parts of Tcl's sampleextension and started
	development of libmagic interface ***


generic/tmag.c
/* --------------------------------------------------------------------------
 * tmag.c --
 *
 *	This file implements a Tcl interface to the libmagic functions.
 *
 * Copyright © 2008-2012 Matthias Kraft <M.Kraft@gmx.com>.
 *
 * See the file "license.terms" for information on usage and redistribution
 * of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 *
 * -------------------------------------------------------------------------- */

/* string header */
#include <string.h>
/* Tcl header */
#include <tcl.h>
/* tmag header */
#include "tmag.h"

/* --------------------------------------------------------------------------
 * TmagSetErrorResult --
 *
 * Create Tcl objects from strings and set them as interp result.
 * -------------------------------------------------------------------------- */

void TmagSetErrorResult(Tcl_Interp *interp, const char *message, const char *reason) {
  Tcl_Obj *result_obj;

  result_obj = Tcl_NewStringObj(message, -1);
  Tcl_AppendToObj(result_obj, reason, -1);
  Tcl_SetObjResult(interp, result_obj);
}

/* --------------------------------------------------------------------------
 * TmagSessionInit --
 *
 * Initialize a new libmagic session and load the default database.
 * -------------------------------------------------------------------------- */

magic_t TmagSessionInit(Tcl_Interp *interp, int flags) {
  magic_t magic_h;

  if ((magic_h = magic_open(TMAG_STANDARD_FLAGS|flags)) == NULL) {
    TmagSetErrorResult(interp, TMAG_OPEN_ERROR_MSG, magic_error(magic_h));

  } else if ((magic_load(magic_h, NULL)) != 0) {
    TmagSetErrorResult(interp, TMAG_LOAD_ERROR_MSG, magic_error(magic_h));
    magic_close(magic_h);
    magic_h = NULL;
  }

  return magic_h;
}

/* --------------------------------------------------------------------------
 * TmagFileCmd --
 *
 * Implements Tcl command
 * libmagic::filetype filename ?(-isbuffer|-follow)? ?-all?
 *
 * Returns result string from libmagic or an error.
 * -------------------------------------------------------------------------- */

int TmagFileCmd(ClientData data, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]) {
    tmagData *session = (tmagData *)data;
    const char *result_str;
    CONST char *fname_str=NULL;
    const unsigned char *buffer_ptr=NULL;
    int buffer_len=0, rc=TCL_OK, flag_isbuffer=0, flag_follow=0, magic_flags=TMAG_STANDARD_FLAGS;

    /* possible flags */
    static const char *flags[] = {"-isbuffer", "-follow", "-all", "-raw", NULL};
    enum flagsIdx {IsBufferIdx, FollowIdx, AllIdx, RawIdx};

    /* check possible range of arguments and process them,
     * note that -isbuffer and -follow are mutual exclusive */
    if (objc < 2 || objc > 4) {
	Tcl_WrongNumArgs(interp, 1, objv, "filename ?flags ...?");
	return TCL_ERROR;
    }
    while (objc > 2) {
      int idx;
      if (Tcl_GetIndexFromObj(interp, objv[objc-1], flags, "flag", 0, &idx) != TCL_OK) {
	return TCL_ERROR;
      }
      switch (idx) {
      case IsBufferIdx:
	flag_isbuffer = 1;
	break;
      case FollowIdx:
	flag_follow = 1;
	magic_flags |= MAGIC_SYMLINK;
	break;
      case AllIdx:
	magic_flags |= MAGIC_CONTINUE;
	break;
      case RawIdx:
	magic_flags |= MAGIC_RAW;
	break;
      }
      objc--;
    }
    if (flag_follow && flag_isbuffer) {
      TmagSetErrorResult(interp, TMAG_ILL_FLAGS_MSG, TMAG_ILL_FLAGS_RSN);
      return TCL_ERROR;
    }
    if (flag_isbuffer) {
      buffer_ptr = Tcl_GetByteArrayFromObj(objv[1], &buffer_len);
    } else {
      fname_str = Tcl_GetString(objv[1]);
    }

    /* now that we have all required input, configure libmagic */
    if ((magic_setflags(session->magic_h, magic_flags)) < 0) {
      TmagSetErrorResult(interp, TMAG_FLAGS_ERROR_MSG, magic_error(session->magic_h));
      return TCL_ERROR;
    }

    /* identify content */
    if (flag_isbuffer) {
      result_str = magic_buffer(session->magic_h, (const void *)buffer_ptr, buffer_len);
    } else {
      result_str = magic_file(session->magic_h, fname_str);
    }

    /* check for error, setup result, close libmagic session */
    if (result_str == NULL) {
      int eno;
      if ((eno = magic_errno(session->magic_h)) == 0) {
	TmagSetErrorResult(interp, (flag_isbuffer ? TMAG_BUFFER_ERROR_MSG : TMAG_FILE_ERROR_MSG), magic_error(session->magic_h) );
      } else {
	TmagSetErrorResult(interp, (flag_isbuffer ? TMAG_BUFFER_ERROR_MSG : TMAG_FILE_ERROR_MSG), strerror(eno) );
      }
      rc = TCL_ERROR;
    } else {
      Tcl_SetObjResult(interp, Tcl_NewStringObj(result_str, -1));
    }
    return rc;
}

/* --------------------------------------------------------------------------
 * TmagMimeCmd --
 *
 * Implements Tcl command
 * libmagic::mimetype filename ?(-isbuffer|-follow)? ?-all? ?-with-encoding?
 *
 * Returns the mime type as result string, with a string identifying the
 * encoding optionally appended. May return an error.
 * -------------------------------------------------------------------------- */

int TmagMimeCmd(ClientData data, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]) {
    tmagData *session = (tmagData *)data;
    const char *result_str;
    CONST char *fname_str=NULL;
    const unsigned char *buffer_ptr=NULL;
    int buffer_len=0, rc=TCL_OK, flag_isbuffer=0, flag_follow=0, magic_flags=(TMAG_STANDARD_FLAGS|MAGIC_MIME_TYPE);

    /* possible flags */
    static const char *flags[] = {"-isbuffer", "-follow", "-all", "-with-encoding", NULL};
    enum flagsIdx {IsBufferIdx, FollowIdx, AllIdx, WithEncodingIdx};

    /* check possible range of arguments and process them,
     * note that -isbuffer and -follow are mutual exclusive */
    if (objc < 2 || objc > 5) {
	Tcl_WrongNumArgs(interp, 1, objv, "filename ?flags ...?");
	return TCL_ERROR;
    }
    while (objc > 2) {
      int idx;
      if (Tcl_GetIndexFromObj(interp, objv[objc-1], flags, "flag", 0, &idx) != TCL_OK) {
	return TCL_ERROR;
      }
      switch (idx) {
      case IsBufferIdx:
	flag_isbuffer = 1;
	break;
      case FollowIdx:
	flag_follow = 1;
	magic_flags |= MAGIC_SYMLINK;
	break;
      case AllIdx:
	magic_flags |= MAGIC_CONTINUE;
	break;
      case WithEncodingIdx:
	magic_flags |= MAGIC_MIME_ENCODING;
	break;
      }
      objc--;
    }
    if (flag_follow && flag_isbuffer) {
      TmagSetErrorResult(interp, TMAG_ILL_FLAGS_MSG, TMAG_ILL_FLAGS_RSN);
      return TCL_ERROR;
    }
    if (flag_isbuffer) {
      buffer_ptr = Tcl_GetByteArrayFromObj(objv[1], &buffer_len);
    } else {
      fname_str = Tcl_GetString(objv[1]);
    }

    /* now that we have all required input, configure libmagic */
    if ((magic_setflags(session->magic_h, magic_flags)) < 0) {
      TmagSetErrorResult(interp, TMAG_FLAGS_ERROR_MSG, magic_error(session->magic_h));
      return TCL_ERROR;
    }

    /* identify content */
    if (flag_isbuffer) {
      result_str = magic_buffer(session->magic_h, (const void *)buffer_ptr, buffer_len);
    } else {
      result_str = magic_file(session->magic_h, fname_str);
    }

    /* check for error, setup result, close libmagic session */
    if (result_str == NULL) {
      int eno;
      if ((eno = magic_errno(session->magic_h)) == 0) {
	TmagSetErrorResult(interp, (flag_isbuffer ? TMAG_BUFFER_ERROR_MSG : TMAG_FILE_ERROR_MSG), magic_error(session->magic_h) );
      } else {
	TmagSetErrorResult(interp, (flag_isbuffer ? TMAG_BUFFER_ERROR_MSG : TMAG_FILE_ERROR_MSG), strerror(eno) );
      }
      rc = TCL_ERROR;
    } else {
      Tcl_SetObjResult(interp, Tcl_NewStringObj(result_str, -1));
    }
    return rc;
}

/* --------------------------------------------------------------------------
 * Tmag_Init --
 *
 * Register namespace, commands, and package with the Tcl interpreter.
 * -------------------------------------------------------------------------- */

int Tmag_Init(Tcl_Interp *interp) {
    Tcl_Namespace *nsPtr; /* pointer to hold our own new namespace */
    tmagData *session;

    if (Tcl_InitStubs(interp, TCL_VERSION, 0) == NULL) {
        return TCL_ERROR;
    }

    if ((session = (tmagData *)Tcl_Alloc(sizeof(tmagData))) == NULL) {
      return TCL_ERROR;
    }

    /* now that we have all required input, init libmagic */
    if ((session->magic_h = TmagSessionInit(interp, 0)) == NULL) {
      Tcl_Free((char *)session);
      return TCL_ERROR;
    }

    /* create the namespace */
    if ((nsPtr = Tcl_CreateNamespace(interp, TMAG_NS, NULL, NULL)) == NULL) {
      magic_close(session->magic_h);
      Tcl_Free((char *)session);
      return TCL_ERROR;
    }

    Tcl_CreateObjCommand(interp, TMAG_NS "::filetype", TmagFileCmd, (ClientData)session, NULL);
    Tcl_CreateObjCommand(interp, TMAG_NS "::mimetype", TmagMimeCmd, (ClientData)session, NULL);

    if (Tcl_PkgProvide(interp, TMAG_EXT_NAME, TMAG_EXT_VERSION) == TCL_ERROR) {
      /* close leaks */
        return TCL_ERROR;
    }

    return TCL_OK;
}

/* TODO: close leaks:
 * - magic_close(session->magic_h)
 * - Tcl_Free(session)
 */
#if 10 * TCL_MAJOR_VERSION + TCL_MINOR_VERSION >= 85
int Tmag_Unload(Tcl_Interp *interp, int flags) {
  /* TODO: how do I get my hands on ClientData now? */
  /* check Tcl_(G|S)etAssocData() */
  TmagSetErrorResult(interp, TMAG_EXT_NAME, TMAG_UNLOAD_ERROR_MSG);
  return TCL_ERROR;
}

#endif /* Tcl 8.5 */


generic/tmag.h
/* --------------------------------------------------------------------------
 * tmag.h --
 *
 *	This file contains definitions for the implemention of a
 *	Tcl interface to the libmagic functions.
 *
 * Copyright © 2008-2012 Matthias Kraft <M.Kraft@gmx.com>.
 *
 * See the file "license.terms" for information on usage and redistribution
 * of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 * -------------------------------------------------------------------------- */

#ifndef TMAG_H
#define TMAG_H

/* libmagic header */
#include <magic.h>

/* package definition, from -D options*/
#define TMAG_EXT_NAME		PACKAGE_NAME
#define TMAG_EXT_VERSION	PACKAGE_VERSION

/* namespace definition */
#define TMAG_NS			"libmagic"

/* error message definitions */
#define TMAG_OPEN_ERROR_MSG	"inititializing new libmagic session failed: "
#define TMAG_LOAD_ERROR_MSG	"loading the default libmagic database failed: "
#define TMAG_BUFFER_ERROR_MSG	"buffer type identification failed: "
#define TMAG_FILE_ERROR_MSG	"file type identification failed: "
#define TMAG_FLAGS_ERROR_MSG	"configuring libmagic query: "
#define TMAG_UNLOAD_ERROR_MSG	" unload is not supported"

#define TMAG_ILL_FLAGS_MSG	"illegal combination of flags: "
#define TMAG_ILL_FLAGS_RSN	"cannot resolve symlinks on buffer input"

/* data definitions */
typedef struct tmagData {
  magic_t magic_h;
} tmagData;

#define TMAG_STANDARD_FLAGS	(MAGIC_ERROR)

#endif /* TMAG_H */


tmag.tcl
# --------------------------------------------------------------------------
# -- tmag.tcl
#
#	This file implements a Tcl interface to the libmagic functions.
#
# Copyright © 2008-2009 Matthias Kraft <M.Kraft@gmx.com>.
#
# See the file "license.terms" for information on usage and redistribution
# of this file, and for a DISCLAIMER OF ALL WARRANTIES.
#
# --------------------------------------------------------------------------

package require critcl

# Header files: string header, Tcl header, tmag header
# FIXME: Tcl header required? better use critcl::cheaders for tmag.h?
::critcl::ccode {
#include <string.h>
#include <tcl.h>
#include "tmag.h"
}

# --------------------------------------------------------------------------
# TmagSetErrorResult --
#
# Create Tcl objects from strings and set them as interp result.
# --------------------------------------------------------------------------

::critcl::ccode {
void TmagSetErrorResult(Tcl_Interp *interp, const char *message, const char *reason) {
  Tcl_Obj *result_obj;

  result_obj = Tcl_NewStringObj(message, -1);
  Tcl_AppendToObj(result_obj, reason, -1);
  Tcl_SetObjResult(interp, result_obj);
}
}

# --------------------------------------------------------------------------
# TmagSessionInit --
#
# Initialize a new libmagic session and load the default database.
# --------------------------------------------------------------------------

::critcl::ccode {
magic_t TmagSessionInit(Tcl_Interp *interp, int flags) {
  magic_t magic_h;

  if ((magic_h = magic_open(TMAG_STANDARD_FLAGS|flags)) == NULL) {
    TmagSetErrorResult(interp, TMAG_OPEN_ERROR_MSG, magic_error(magic_h));

  } else if ((magic_load(magic_h, NULL)) != 0) {
    TmagSetErrorResult(interp, TMAG_LOAD_ERROR_MSG, magic_error(magic_h));
    magic_close(magic_h);
    magic_h = NULL;
  }

  return magic_h;
}
}

# --------------------------------------------------------------------------
# TmagFileCmd --
#
# Implements Tcl command
# libmagic::filetype filename ?(-isbuffer|-follow)? ?-all?
#
# Returns result string from libmagic or an error.
# --------------------------------------------------------------------------
# FIXME: change to ::critcl::cproc!

::critcl::ccode {
int TmagFileCmd(ClientData data, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]) {
    tmagData *session = (tmagData *)data;
    const char *result_str;
    CONST char *fname_str=NULL;
    const unsigned char *buffer_ptr=NULL;
    int buffer_len=0, rc=TCL_OK, flag_isbuffer=0, flag_follow=0, magic_flags=TMAG_STANDARD_FLAGS;

    /* possible flags */
    const char *flags[] = {"-isbuffer", "-follow", "-all", "-raw", NULL};
    enum flagsIdx {IsBufferIdx, FollowIdx, AllIdx, RawIdx};

    /* check possible range of arguments and process them,
     * note that -isbuffer and -follow are mutual exclusive */
    if (objc < 2 || objc > 4) {
	Tcl_WrongNumArgs(interp, 1, objv, "filename ?flags ...?");
	return TCL_ERROR;
    }
    while (objc > 2) {
      int idx;
      if (Tcl_GetIndexFromObj(interp, objv[objc-1], flags, "flag", 0, &idx) != TCL_OK) {
	return TCL_ERROR;
      }
      switch (idx) {
      case IsBufferIdx:
	flag_isbuffer = 1;
	break;
      case FollowIdx:
	flag_follow = 1;
	magic_flags |= MAGIC_SYMLINK;
	break;
      case AllIdx:
	magic_flags |= MAGIC_CONTINUE;
	break;
      case RawIdx:
	magic_flags |= MAGIC_RAW;
      }
      objc--;
    }
    if (flag_follow && flag_isbuffer) {
      TmagSetErrorResult(interp, TMAG_ILL_FLAGS_MSG, TMAG_ILL_FLAGS_RSN);
      return TCL_ERROR;
    }
    if (flag_isbuffer) {
      buffer_ptr = Tcl_GetByteArrayFromObj(objv[1], &buffer_len);
    } else {
      fname_str = Tcl_GetString(objv[1]);
    }

    /* now that we have all required input, configure libmagic */
    if ((magic_setflags(session->magic_h, magic_flags)) < 0) {
      TmagSetErrorResult(interp, TMAG_FLAGS_ERROR_MSG, magic_error(session->magic_h));
      return TCL_ERROR;
    }

    /* identify content */
    if (flag_isbuffer) {
      result_str = magic_buffer(session->magic_h, (const void *)buffer_ptr, buffer_len);
    } else {
      result_str = magic_file(session->magic_h, fname_str);
    }

    /* check for error, setup result, close libmagic session */
    if (result_str == NULL) {
      int eno;
      if ((eno = magic_errno(session->magic_h)) == 0) {
	TmagSetErrorResult(interp, (flag_isbuffer ? TMAG_BUFFER_ERROR_MSG : TMAG_FILE_ERROR_MSG), magic_error(session->magic_h) );
      } else {
	TmagSetErrorResult(interp, (flag_isbuffer ? TMAG_BUFFER_ERROR_MSG : TMAG_FILE_ERROR_MSG), strerror(eno) );
      }
      rc = TCL_ERROR;
    } else {
      Tcl_SetObjResult(interp, Tcl_NewStringObj(result_str, -1));
    }
    return rc;
}
}

# --------------------------------------------------------------------------
# TmagMimeCmd --
#
# Implements Tcl command
# libmagic::mimetype filename ?(-isbuffer|-follow)? ?-all? ?-with-encoding?
#
# Returns the mime type as result string, with a string identifying the
# encoding optionally appended. May return an error.
# --------------------------------------------------------------------------
# FIXME: change to critcl::cproc!
::critcl::ccode {
int TmagMimeCmd(ClientData data, Tcl_Interp *interp, int objc, Tcl_Obj *CONST objv[]) {
    tmagData *session = (tmagData *)data;
    const char *result_str;
    CONST char *fname_str=NULL;
    const unsigned char *buffer_ptr=NULL;
    int buffer_len=0, rc=TCL_OK, flag_isbuffer=0, flag_follow=0, magic_flags=(TMAG_STANDARD_FLAGS|MAGIC_MIME_TYPE);

    /* possible flags */
    const char *flags[] = {"-isbuffer", "-follow", "-all", "-with-encoding", NULL};
    enum flagsIdx {IsBufferIdx, FollowIdx, AllIdx, WithEncodingIdx};

    /* check possible range of arguments and process them,
     * note that -isbuffer and -follow are mutual exclusive */
    if (objc < 2 || objc > 5) {
	Tcl_WrongNumArgs(interp, 1, objv, "filename ?flags ...?");
	return TCL_ERROR;
    }
    while (objc > 2) {
      int idx;
      if (Tcl_GetIndexFromObj(interp, objv[objc-1], flags, "flag", 0, &idx) != TCL_OK) {
	return TCL_ERROR;
      }
      switch (idx) {
      case IsBufferIdx:
	flag_isbuffer = 1;
	break;
      case FollowIdx:
	flag_follow = 1;
	magic_flags |= MAGIC_SYMLINK;
	break;
      case AllIdx:
	magic_flags |= MAGIC_CONTINUE;
	break;
      case WithEncodingIdx:
	magic_flags |= MAGIC_MIME_ENCODING;
	break;
      }
      objc--;
    }
    if (flag_follow && flag_isbuffer) {
      TmagSetErrorResult(interp, TMAG_ILL_FLAGS_MSG, TMAG_ILL_FLAGS_RSN);
      return TCL_ERROR;
    }
    if (flag_isbuffer) {
      buffer_ptr = Tcl_GetByteArrayFromObj(objv[1], &buffer_len);
    } else {
      fname_str = Tcl_GetString(objv[1]);
    }

    /* now that we have all required input, configure libmagic */
    if ((magic_setflags(session->magic_h, magic_flags)) < 0) {
      TmagSetErrorResult(interp, TMAG_FLAGS_ERROR_MSG, magic_error(session->magic_h));
      return TCL_ERROR;
    }

    /* identify content */
    if (flag_isbuffer) {
      result_str = magic_buffer(session->magic_h, (const void *)buffer_ptr, buffer_len);
    } else {
      result_str = magic_file(session->magic_h, fname_str);
    }

    /* check for error, setup result, close libmagic session */
    if (result_str == NULL) {
      int eno;
      if ((eno = magic_errno(session->magic_h)) == 0) {
	TmagSetErrorResult(interp, (flag_isbuffer ? TMAG_BUFFER_ERROR_MSG : TMAG_FILE_ERROR_MSG), magic_error(session->magic_h) );
      } else {
	TmagSetErrorResult(interp, (flag_isbuffer ? TMAG_BUFFER_ERROR_MSG : TMAG_FILE_ERROR_MSG), strerror(eno) );
      }
      rc = TCL_ERROR;
    } else {
      Tcl_SetObjResult(interp, Tcl_NewStringObj(result_str, -1));
    }
    return rc;
}
}

# --------------------------------------------------------------------------
# Tmag_Init --
#
# Register namespace, commands, and package with the Tcl interpreter.
# --------------------------------------------------------------------------
# FIXME: put into critcl::cinit?
::critcl::ccode {
int Tmag_Init(Tcl_Interp *interp) {
    Tcl_Namespace *nsPtr; /* pointer to hold our own new namespace */
    tmagData *session;

    if (Tcl_InitStubs(interp, TCL_VERSION, 0) == NULL) {
        return TCL_ERROR;
    }

    if ((session = (tmagData *)Tcl_Alloc(sizeof(tmagData))) == NULL) {
      return TCL_ERROR;
    }

    /* now that we have all required input, init libmagic */
    if ((session->magic_h = TmagSessionInit(interp, 0)) == NULL) {
      Tcl_Free(session);
      return TCL_ERROR;
    }

    /* create the namespace */
    if ((nsPtr = Tcl_CreateNamespace(interp, TMAG_NS, NULL, NULL)) == NULL) {
      magic_close(session->magic_h);
      Tcl_Free(session);
      return TCL_ERROR;
    }

    Tcl_CreateObjCommand(interp, TMAG_NS "::filetype", TmagFileCmd, (ClientData)session, NULL);
    Tcl_CreateObjCommand(interp, TMAG_NS "::mimetype", TmagMimeCmd, (ClientData)session, NULL);

    if (Tcl_PkgProvide(interp, TMAG_EXT_NAME, TMAG_EXT_VERSION) == TCL_ERROR) {
      /* close leaks */
        return TCL_ERROR;
    }

    return TCL_OK;
}
}

# close leak:
# - magic_close(session->magic_h)
# - Tcl_Free(session)
# FIXME: how to handle the unload stuff with critcl?

::critcl::ccode {
#if 10 * TCL_MAJOR_VERSION + TCL_MINOR_VERSION >= 85
int Tmag_Unload(Tcl_Interp *interp, int flags) {
  /* how do I get my hands on ClientData now? */
  /* check Tcl_(G|S)etAssocData() */
}
#endif /* Tcl 8.5 */
}
